cupy/cuda/memory_hook.py

```diff
@@ -0,0 +1,160 @@
+import collections
+import threading
+
+thread_local = threading.local()
+
+
+def get_memory_hooks():
+    if not hasattr(thread_local, 'memory_hooks'):
+        thread_local.memory_hooks = collections.OrderedDict()
+    return thread_local.memory_hooks
+
+
+class MemoryHook(object):
+    """Base class of hooks for Memory allocations.
+
+    :class:`~cupy.cuda.MemoryHook` is an callback object
+    that is registered to :class:`~cupy.cuda.SingleDeviceMemoryPool`.
+    Registered memory hooks are invoked before and after
+    memory is allocated from GPU device, and
+    memory is retrieved from memory pool, and
+    memory is released to memory pool.
+
+    Memory hooks that derive :class:`MemoryHook` are required
+    to implement six methods:
+    :meth:`~cupy.cuda.MemoryHook.alloc_preprocess`,
+    :meth:`~cupy.cuda.MemoryHook.alloc_postprocess`,
+    :meth:`~cupy.cuda.MemoryHook.malloc_preprocess`,
+    :meth:`~cupy.cuda.MemoryHook.malloc_postprocess`,
+    :meth:`~cupy.cuda.MemoryHook.free_preprocess`, and
+    :meth:`~cupy.cuda.MemoryHook.free_postprocess`,
+    By default, these methods do nothing.
+
+    Specifically, :meth:`~cupy.cuda.MemoryHook.alloc_preprocess`
+    (resp. :meth:`~cupy.cuda.MemoryHook.alloc_postprocess`)
+    of all memory hooks registered are called before (resp. after)
+    memory is allocated from GPU device.
+
+    Likewise, :meth:`~cupy.cuda.MemoryHook.malloc_preprocess`
+    (resp. :meth:`~cupy.cuda.MemoryHook.malloc_postprocess`)
+    of all memory hooks registered are called before (resp. after)
+    memory is retrieved from memory pool, that is,
+    :meth:`~cupy.cuda.SingleDeviceMemoryPool.malloc` is invoked.
+
+    Below is the pseudo code to descirbe how
+    :meth:`~cupy.cuda.SingleDeviceMemoryPool.malloc` and hooks work.
+    Please note that ``alloc_preprocess`` and ``alloc_postprocess``
+    are not invoked if a cached free chunk is found.
+
+        def malloc(size):
```

Comment:
Yes, that is what I meant. I understand that the code block is not appropriate for pseudo code. But I tried to registered the document to the docstring and noticed that current pseudo code is not well-rendered (see the following screenshot).

<img width="954" alt="2017-08-10 19 26 50" src="https://user-images.githubusercontent.com/407052/29166648-3afa3216-7e02-11e7-9d01-8a4e53f5cc40.png">

If there would be a standard way to format pseudo code, I would like to use it.